Fix crash while generating planning percentile func on partition tables

bhuvnesh2703 opened the pull request:
Fix crash while generarting multistage agg plan using percentile func on partition tables

Function within_agg_make_baseplan() invokes choose_deduplicate() which creates a
flattened rte with the assumption that it is generated later in
query_planner and releases it before returning.
However,  within_agg_make_baseplan() takes care of releasing the memory
during deduplicate optimization and it crashes while trying to release it again.

Also, before this function is invoked, query_planner has already
generated the flattened rte as it is not going to change, so we need not
generate it again in choose_deduplicate() as its already available.

In GPDB4, choose_deduplicate() method does not has the code removed in
this commit as well and the issue reported in 5 does not exists in
GPDB4. In GPDB Master, this portion of the code has significantly
changed due to merge from upstream and choose_deduplicate() method does
not exist nor the issue.

This code was earlier added during postgres 8.3 merge via commit a453004.

Adding tests to validate the change.

d commented:
1. The JIRA number is wrong
1. It's not immediately clear from the test, nor commit message (nor the code comment) what was broken before this patch
1. The wording here seems to imply a crash (that's my best uneducated guess) that we're fixing. However, running `percentile.sql` in this patch without the code does *not* induce a crash. For context, I got this as the error:

```
-- MPP-29322: test planner multistage agg plan for agg function median on partitioned table.
-- orca falls back for median function
DROP TABLE IF EXISTS t_multistage;
psql:sql/percentile.sql:280: NOTICE:  table "t_multistage" does not exist, skipping
DROP TABLE
CREATE TABLE t_multistage (a int, b int)
PARTITION BY RANGE(b) (START(1) END (15) EVERY(1));
psql:sql/percentile.sql:282: NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
psql:sql/percentile.sql:282: NOTICE:  CREATE TABLE will create partition "t_multistage_1_prt_1" for table "t_multistage"
psql:sql/percentile.sql:282: NOTICE:  CREATE TABLE will create partition "t_multistage_1_prt_2" for table "t_multistage"
psql:sql/percentile.sql:282: NOTICE:  CREATE TABLE will create partition "t_multistage_1_prt_3" for table "t_multistage"
psql:sql/percentile.sql:282: NOTICE:  CREATE TABLE will create partition "t_multistage_1_prt_4" for table "t_multistage"
psql:sql/percentile.sql:282: NOTICE:  CREATE TABLE will create partition "t_multistage_1_prt_5" for table "t_multistage"
psql:sql/percentile.sql:282: NOTICE:  CREATE TABLE will create partition "t_multistage_1_prt_6" for table "t_multistage"
psql:sql/percentile.sql:282: NOTICE:  CREATE TABLE will create partition "t_multistage_1_prt_7" for table "t_multistage"
psql:sql/percentile.sql:282: NOTICE:  CREATE TABLE will create partition "t_multistage_1_prt_8" for table "t_multistage"
psql:sql/percentile.sql:282: NOTICE:  CREATE TABLE will create partition "t_multistage_1_prt_9" for table "t_multistage"
psql:sql/percentile.sql:282: NOTICE:  CREATE TABLE will create partition "t_multistage_1_prt_10" for table "t_multistage"
psql:sql/percentile.sql:282: NOTICE:  CREATE TABLE will create partition "t_multistage_1_prt_11" for table "t_multistage"
psql:sql/percentile.sql:282: NOTICE:  CREATE TABLE will create partition "t_multistage_1_prt_12" for table "t_multistage"
psql:sql/percentile.sql:282: NOTICE:  CREATE TABLE will create partition "t_multistage_1_prt_13" for table "t_multistage"
psql:sql/percentile.sql:282: NOTICE:  CREATE TABLE will create partition "t_multistage_1_prt_14" for table "t_multistage"
CREATE TABLE
INSERT INTO t_multistage SELECT i, (i%14)+1 FROM generate_series(1,100)i;
INSERT 0 100
EXPLAIN SELECT median(b) from t_multistage;
psql:sql/percentile.sql:284: ERROR:  unrecognized node type: 0 (walkers.c:535)
SELECT median(b) from t_multistage;
psql:sql/percentile.sql:285: ERROR:  unrecognized node type: 0 (walkers.c:535)
DROP TABLE IF EXISTS t_multistage;
DROP TABLE

```

asimrp commented on src/test/regress/sql/percentile.sql:
```diff
@@ -275,6 +275,16 @@ from  mpp_22413
 where d2 ='55'
 group by d1, d2;
 
+-- MPP-29322: test planner multistage agg plan for agg function median on partitioned table.
```
Can we please avoid introducing internal jira numbers?  Some of us are diligently removing existing jira numbers by describing the issue more elaborately, so that the code remains self-contained.

bhuvnesh2703 replied:
sure.. point taken..

d commented:
To clarify: I have zero opinion on whether we should use JIRA numbers *as long as it helps me in the future*. And a wrong JIRA number is a negative.

bhuvnesh2703 commented:
@d Thanks for pointing out.. Based on the number of segments and number of partitions, the issue was observed via two different symptoms. I believe when i created the repro i had a 1 segment cluster, but anyways i have added the crash repro for 3 node cluster and removed the JIRA number.

```
CREATE TABLE t_multistage (a int, b int) 
PARTITION BY RANGE(b) (START(1) END (15) EVERY(1));
pivotal=# SELECT median(b) from t_multistage;
ERROR:  unrecognized node type: 0 (walkers.c:535)

pivotal=# CREATE TABLE t_multistage (a int, b int)
PARTITION BY RANGE(b) (START(1) END (18) EVERY(1));
pivotal=# SELECT median(b) from t_multistage;server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
The connection to the server was lost. Attempting reset: Failed.
```

d commented:
I don't understand why `18` crashes but `15` didn't. But I'll leave that to the reviewers of this PR...

bhuvnesh2703 commented:
@d as we spoke, the problem here occurs with 15 and not with 18 is due to the reuse of memory allocated in case of 15 vs 18. So with the pfree(root->simple_rte_array) (removed in this PR) we set the memory location of `root->simple_rte_array` elements to point to 0x7f, which is all good, but later we do another allocation for Flow node in `mark_plan_strewn(result_plan);`. In case of 15 partition tables, the allocated memory for flow node created in `mark_plan_strewn` is taken from the free chunklist to which simple_rte_array was pointing, but in case of 18 partition table it lies in a different pool. Thus different behavior observed.
